File: backend/src/config/services/clientService.ts
import { pool } from '../database.js';

export const findAll = async () => {
  const result = await pool.query('SELECT * FROM clients ORDER BY id ASC');
  return result.rows;
};

export const findById = async (id: string) => {
  const result = await pool.query('SELECT * FROM clients WHERE id = $1', [id]);
  return result.rows[0] ?? null;
};

export const create = async (data: { name: string; email?: string; phone?: string; address?: string; vat_number?: string; notes?: string }) => {
  const result = await pool.query(
    'INSERT INTO clients (name, email, phone, address, vat_number, notes) VALUES ($1, $2, $3, $4, $5, $6) RETURNING *',
    [data.name, data.email, data.phone, data.address, data.vat_number, data.notes]
  );
  return result.rows[0];
};

export const update = async (id: string, data: { name: string; email?: string; phone?: string; address?: string; vat_number?: string; notes?: string }) => {
  const result = await pool.query(
    'UPDATE clients SET name = $1, email = $2, phone = $3, address = $4, vat_number = $5, notes = $6, updated_at = NOW() WHERE id = $7 RETURNING *',
    [data.name, data.email, data.phone, data.address, data.vat_number, data.notes, id]
  );
  return result.rows[0] ?? null;
};

export const remove = async (id: string) => {
  const result = await pool.query('DELETE FROM clients WHERE id = $1 RETURNING id', [id]);
  return (result.rowCount ?? 0) > 0;
};

File: backend/services/client_service.go
package services

import (
	"aura-erp/backend/config"
	"aura-erp/backend/models"
)

func GetAllClients() ([]models.Client, error) {
	query := `SELECT id, name, email, phone, address, vat_number, notes, created_at, updated_at FROM clients ORDER BY id ASC`
	rows, err := config.DB.Query(query)
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	var clients []models.Client
	for rows.Next() {
		var client models.Client
		if err := rows.Scan(&client.ID, &client.Name, &client.Email, &client.Phone, &client.Address, &client.VatNumber, &client.Notes, &client.CreatedAt, &client.UpdatedAt); err != nil {
			return nil, err
		}
		clients = append(clients, client)
	}

	return clients, nil
}

func GetClientByID(id string) (*models.Client, error) {
	query := `SELECT id, name, email, phone, address, vat_number, notes, created_at, updated_at FROM clients WHERE id = $1`

	var client models.Client
	err := config.DB.QueryRow(query, id).Scan(
		&client.ID, &client.Name, &client.Email, &client.Phone, &client.Address, &client.VatNumber, &client.Notes, &client.CreatedAt, &client.UpdatedAt,
	)
	if err != nil {
		return nil, err
	}

	return &client, nil
}

func CreateClient(data models.ClientCreate) (*models.Client, error) {
	query := `INSERT INTO clients (name, email, phone, address, vat_number, notes)
	          VALUES ($1, $2, $3, $4, $5, $6)
	          RETURNING id, name, email, phone, address, vat_number, notes, created_at, updated_at`

	var client models.Client
	err := config.DB.QueryRow(query, data.Name, data.Email, data.Phone, data.Address, data.VatNumber, data.Notes).Scan(
		&client.ID, &client.Name, &client.Email, &client.Phone, &client.Address, &client.VatNumber, &client.Notes, &client.CreatedAt, &client.UpdatedAt,
	)
	if err != nil {
		return nil, err
	}

	return &client, nil
}

func UpdateClient(id string, data models.ClientUpdate) (*models.Client, error) {
	query := `UPDATE clients SET name = $1, email = $2, phone = $3, address = $4, vat_number = $5, notes = $6, updated_at = NOW()
	          WHERE id = $7
	          RETURNING id, name, email, phone, address, vat_number, notes, created_at, updated_at`

	var client models.Client
	err := config.DB.QueryRow(query, data.Name, data.Email, data.Phone, data.Address, data.VatNumber, data.Notes, id).Scan(
		&client.ID, &client.Name, &client.Email, &client.Phone, &client.Address, &client.VatNumber, &client.Notes, &client.CreatedAt, &client.UpdatedAt,
	)
	if err != nil {
		return nil, err
	}

	return &client, nil
}

func DeleteClient(id string) error {
	query := `DELETE FROM clients WHERE id = $1`
	result, err := config.DB.Exec(query, id)
	if err != nil {
		return err
	}

	rowsAffected, err := result.RowsAffected()
	if err != nil || rowsAffected == 0 {
		return err
	}

	return nil
}
